### PR TITLE
Fix random number seeding

### PR DESCRIPTION
**Pull request by bowd**
While working on linking the random numbers to quanlib @sissnad identified some issues with the RNG logic. This PR addresses two major issues:

1. **Random number seeds were not consistent between runs**: This was due to the builtin `hash` function not being consistent across runs. I replaced that with a slightly convoluted usage of `hashlib.md5` to extract a truncated int from it as the seed generators use ints in the seed sequence.
2. **Different monte carlo runs were not affecting the RNGs seed sequence**: In order to fix this I opted to move from the current singleton approach from the RNGProvider to an approach where I inject it via the `Engine` as part of the params. This way we can control how it's initialized for each run according to the `run_index` and include that as part of the seed sequence under the hood. This way different montecarlo runs will generate different seed sequences.

### CODE REVIEW

File: model/generators/oracles.py [hunk 1]
```diff
@@ -27,13 +27,17 @@ class OracleRateGenerator(Generator):
     oracles_by_id: Dict[UUID, OracleProvider]
     oracles_by_pair: Dict[Pair, List[OracleProvider]]
     oracle_pairs: List[Pair]
+    rngp: RNGProvider
+    rng: np.random.Generator
```
Comment (tobikuhlmann) [inline model/generators/oracles.py:31]: is this still used? Don't find usage

File: model/utils/rng_provider.py [hunk 2]
```diff
@@ -5,27 +5,35 @@
...
-        return np.random.default_rng(self.__seed__(context))
+        return np.random.default_rng(self.__seed__(list(context)))
 
     def __seed__(self, context: List[Union[str, int]]) -> np.random.SeedSequence:
-        return np.random.SeedSequence(
-            self.entropy,
-            spawn_key=map(abs, map(hash, context))
+        seed_sequence = np.random.SeedSequence(
+            self.seed,
+            spawn_key=map(abs, map(__hash__, [self.monte_carlo_run] + context))
```
Comment (tobikuhlmann) [inline model/utils/rng_provider.py:32]: can numbers from the `__hash__` function get negative? Curious how negative values would be encoded / decoded

**Review by ghost — APPROVED** [review]
yup, works

File: model/generators/oracles.py (hunk 1 above)
Comment (bowd) [reply model/generators/oracles.py:31]: You're right doesn't look like it's actually used in the generator, I deleted it 👍

File: model/utils/rng_provider.py (hunk 2 above)
Comment (bowd) [reply model/utils/rng_provider.py:32]: Good catch, that was leftover from the builtin `hash` which needed an abs, removing 👍

**Review by sissnad — APPROVED** [review]
Looks good to me

**Pull request merged**